experimental/parser/diagnostics_token.go

```diff
@@ -68,34 +62,31 @@ type ErrUnmatched struct {
 }
 
 // OpenClose returns the expected open/close delimiters for this matched pair.
-func (e ErrUnmatched) OpenClose() (string, string) {
+func (e errUnmatched) OpenClose() (string, string) {
```

Comment:
Should I take this as a general bit of style guidance from you? I have typically taken the convention that methods on unexported are exported if they are intended to be called by functions in other files. But I haven't been super consistent... and I think this is an easier convention to follow.

Reply:
Yeah, maybe it's a peculiar thing specific to me 🤷 

I typically default to. making everything unexported, but especially for values that escape this package (like error types). Otherwise, exported functions could be called from outside the package by external code type-asserting to an interface or via reflection. Just seems like better encapsulation to not have to worry about that (this is method is harmless, but it's particularly bad for methods that mutate the value or could be used to break invariants if called incorrectly). This seems particularly important from a backwards-compatibility story and preventing yourself from being "trapped" into maintaining things due to Hyrum's Law.

So the only times I will export things from an unexported type are:
1. When it's required for interop with other packages -- like fields of a struct used with "encoding/json"
2. When we actually want/need the type to implement an interface -- for use within this package or use outside of it (for cases where instances of the unexported type can escape)
3. When the type has its own helper methods and _it would materially improve readability/maintainability_ to distinguish between the methods that other code is expected to use (exported) and those that should be private to the type (unexported).

> if they are intended to be called by functions in other files

But common refactors like moving things between files or breaking up a large file into multiple files can suddenly mean you're breaking this convention. Without anything to enforce this convention (like a linter), it feels brittle and unreliable, so I'd personally just rather not.

Reply:
I am fairly cognizant of unexported method names escaping. I usually pay attention to those.

I do wish there was a linter for this: enforce that all unexported types' methods are unexported, unless that type is converted into an interface within the package.